File: python/mlx-movie-director/app/lipsync_metrics.py
"""lipsync_metrics — measure whether generated mouth motion tracks the audio.

Answers the open question left by the IA2V vendor-bug fix (see
``docs/openmontage-capability-matrix.md`` "Not yet assessed"): the pipeline
produces a plausible talking-portrait clip, but does mouth motion actually
correlate with speech energy, or is it "a face talking in general"?

Two independent time series are extracted from the same mp4 and correlated:

  - mouth-open ratio per video frame, via mediapipe FaceLandmarker (inner-lip
    gap normalized by interocular distance, so it is invariant to face scale/
    zoom rather than to mouth width, which itself changes during speech).
  - audio RMS envelope, reusing ``app.voice_metrics._frame_rms`` /
    ``app.audio_noise_detect._extract_audio_pcm`` (already-tested decode
    path — no new audio-decoding code).

Both series are resampled onto the video's frame timestamps and compared via
Pearson correlation, searching a small lag window (mouth motion can lag
audio by a frame or two in a joint-diffusion model). The lag-adjusted |r| is
the numeric verdict.
"""

from __future__ import annotations

import os
import subprocess
import tempfile

import numpy as np

from app.audio_noise_detect import _extract_audio_pcm
from app.voice_metrics import _frame_rms, _probe_sample_rate

# Inner-lip vertical gap (top, bottom) and interocular horizontal reference
# (outer eye corners), MediaPipe FaceMesh 468-point topology.
_MOUTH_TOP = 13
_MOUTH_BOTTOM = 14
_EYE_LEFT_OUTER = 33
_EYE_RIGHT_OUTER = 263

# Lag search window in video frames when aligning mouth-open ratio against the
# audio envelope (a few frames covers plausible diffusion-model audio/video
# offset without over-fitting to noise).
_MAX_LAG_FRAMES = 4


def _get_face_landmarker_model_path() -> str:
    """Resolve path to the mediapipe FaceLandmarker task bundle."""
    base = os.path.dirname(os.path.abspath(__file__))
    return os.path.join(base, "models", "face_landmark", "face_landmarker.task")


def _probe_fps(mp4_path: str) -> float:
    """Return the video track's frame rate (fps); 24.0 on any failure."""
    try:
        r = subprocess.run(
            ["ffprobe", "-v", "error", "-select_streams", "v:0",
             "-show_entries", "stream=r_frame_rate", "-of", "csv=p=0", mp4_path],
            capture_output=True, text=True, timeout=30,
        )
        raw = (r.stdout or "").strip()
        if "/" in raw:
            num, den = raw.split("/")
            return float(num) / float(den) if float(den) != 0 else 24.0
        return float(raw) if raw else 24.0
    except Exception:
        return 24.0


def _extract_video_frames(mp4_path: str, out_dir: str) -> list[str]:
    """Extract every video frame as PNG files (frame_0001.png, ...); return sorted paths."""
    pattern = os.path.join(out_dir, "frame_%05d.png")
    cmd = ["ffmpeg", "-i", mp4_path, "-vsync", "0", pattern]
    subprocess.run(cmd, capture_output=True, timeout=60)
    frames = sorted(
        os.path.join(out_dir, f) for f in os.listdir(out_dir) if f.startswith("frame_")
    )
    return frames


def _mouth_open_ratio_from_landmarks(landmarks) -> float | None:
    """Inner-lip gap / interocular distance for one mediapipe landmark list."""
    try:
        top = landmarks[_MOUTH_TOP]
        bottom = landmarks[_MOUTH_BOTTOM]
        eye_l = landmarks[_EYE_LEFT_OUTER]
        eye_r = landmarks[_EYE_RIGHT_OUTER]
    except IndexError:
        return None
    mouth_gap = float(np.hypot(top.x - bottom.x, top.y - bottom.y))
    interocular = float(np.hypot(eye_l.x - eye_r.x, eye_l.y - eye_r.y))
    if interocular <= 1e-9:
        return None
    return mouth_gap / interocular


def extract_mouth_open_series(mp4_path: str) -> dict:
    """Extract per-frame mouth-open ratio for the (single, primary) face in mp4_path.

    Returns:
        dict with keys:
        - ratios: np.ndarray of mouth-open ratios (NaN for frames with no
          detected face)
        - fps: float, video frame rate
        - n_frames: int, total frames processed
        - n_detected: int, frames where a face was found
    """
    try:
        import mediapipe as mp
        from mediapipe.tasks.python import vision
        from mediapipe.tasks.python.core.base_options import BaseOptions
    except ImportError:
        return {"ratios": np.array([]), "fps": 0.0, "n_frames": 0, "n_detected": 0,
                "note": "mediapipe not installed"}

    model_path = _get_face_landmarker_model_path()
    if not os.path.exists(model_path):
        return {"ratios": np.array([]), "fps": 0.0, "n_frames": 0, "n_detected": 0,
                "note": f"FaceLandmarker model not found: {model_path}"}

    fps = _probe_fps(mp4_path)

    with tempfile.TemporaryDirectory() as tmp:
        frame_paths = _extract_video_frames(mp4_path, tmp)
        if not frame_paths:
            return {"ratios": np.array([]), "fps": fps, "n_frames": 0, "n_detected": 0,
                    "note": "no frames extracted"}

        options = vision.FaceLandmarkerOptions(
            base_options=BaseOptions(model_asset_path=model_path),
            num_faces=1,
        )
        landmarker = vision.FaceLandmarker.create_from_options(options)

        ratios = []
        n_detected = 0
        try:
            from PIL import Image
            for frame_path in frame_paths:
                img = Image.open(frame_path).convert("RGB")
                mp_img = mp.Image(image_format=mp.ImageFormat.SRGB, data=np.array(img))
                result = landmarker.detect(mp_img)
                if not result.face_landmarks:
                    ratios.append(float("nan"))
                    continue
                ratio = _mouth_open_ratio_from_landmarks(result.face_landmarks[0])
                if ratio is None:
                    ratios.append(float("nan"))
                else:
                    ratios.append(ratio)
                    n_detected += 1
        finally:
            landmarker.close()

    return {
        "ratios": np.array(ratios, dtype=np.float64),
        "fps": fps,
        "n_frames": len(frame_paths),
        "n_detected": n_detected,
    }


def extract_audio_envelope(mp4_path: str, n_samples: int) -> np.ndarray:
    """Audio RMS envelope resampled onto n_samples points spanning the clip duration."""
    sr = _probe_sample_rate(mp4_path)
    samples = _extract_audio_pcm(mp4_path)
    if len(samples) < 1024 or n_samples <= 0:
        return np.zeros(max(0, n_samples), dtype=np.float64)

    rms_frames = _frame_rms(samples).astype(np.float64)
    if rms_frames.size < 2:
        return np.zeros(n_samples, dtype=np.float64)

    # rms_frames is indexed by audio-frame hop; resample onto n_samples video
    # frame positions spanning the same time range via linear interpolation.
    src_x = np.linspace(0.0, 1.0, rms_frames.size)
    dst_x = np.linspace(0.0, 1.0, n_samples)
    return np.interp(dst_x, src_x, rms_frames)


def _lagged_pearson(a: np.ndarray, b: np.ndarray, max_lag: int) -> tuple[float, int]:
    """Best |Pearson r| between a and b over lags in [-max_lag, max_lag].

    Positive lag means b is shifted forward relative to a (b lags a).
    Returns (best_r, best_lag). NaNs in either series are dropped pairwise
    per-lag before computing r; a lag with fewer than 4 valid pairs is skipped.
    """
    best_r = 0.0
    best_lag = 0
    n = min(len(a), len(b))
    if n < 8:
        return 0.0, 0
    a = a[:n]
    b = b[:n]
    for lag in range(-max_lag, max_lag + 1):
        if lag >= 0:
            a_seg = a[lag:]
            b_seg = b[: n - lag]
        else:
            a_seg = a[: n + lag]
            b_seg = b[-lag:]
        mask = ~(np.isnan(a_seg) | np.isnan(b_seg))
        if mask.sum() < 4:
            continue
        a_valid = a_seg[mask]
        b_valid = b_seg[mask]
        if np.std(a_valid) < 1e-9 or np.std(b_valid) < 1e-9:
            continue
        r = float(np.corrcoef(a_valid, b_valid)[0, 1])
        if abs(r) > abs(best_r):
            best_r = r
            best_lag = lag
    return best_r, best_lag


# Verdict threshold: |r| >= 0.3 is a conventional "weak-to-moderate but real"
# correlation floor in behavioral/signal-correlation literature; below that,
# mouth motion is indistinguishable from audio-independent talking motion.
ADEQUATE_R_THRESHOLD = 0.3


def measure_lipsync_precision(mp4_path: str) -> dict:
    """End-to-end: extract both series from mp4_path, correlate, verdict.

    Returns a dict with mouth/audio series stats, the lag-adjusted Pearson r,
    the chosen lag, and a verdict string ("adequate" | "inadequate" | "no_face"
    | "no_audio").
    """
    mouth = extract_mouth_open_series(mp4_path)
    if mouth.get("n_detected", 0) < 4:
        return {
            "verdict": "no_face",
            "note": mouth.get("note", "insufficient face detections"),
            "n_frames": mouth.get("n_frames", 0),
            "n_detected": mouth.get("n_detected", 0),
        }

    ratios = mouth["ratios"]
    audio_env = extract_audio_envelope(mp4_path, n_samples=len(ratios))
    if np.std(audio_env[~np.isnan(audio_env)] if np.isnan(audio_env).any() else audio_env) < 1e-9:
        return {
            "verdict": "no_audio",
            "note": "audio envelope has ~zero variance (silent/missing track)",
            "n_frames": mouth["n_frames"],
            "n_detected": mouth["n_detected"],
        }

    r, lag = _lagged_pearson(ratios, audio_env, _MAX_LAG_FRAMES)
    # Cross-check with no lag search at all: on short clips, letting the search
    # pick from a shrinking-sample-size window at extreme lags can inflate |r|
    # on noise alone (fewer valid pairs = higher spurious-correlation variance).
    # If the lag-adjusted r and the lag-0 r disagree sharply, the lag search is
    # not to be trusted and the verdict should lean on lag-0.
    mask0 = ~(np.isnan(ratios) | np.isnan(audio_env))
    lag0_r = (
        float(np.corrcoef(ratios[mask0], audio_env[mask0])[0, 1])
        if mask0.sum() >= 4 and np.std(ratios[mask0]) > 1e-9 and np.std(audio_env[mask0]) > 1e-9
        else 0.0
    )
    lag_search_hit_boundary = abs(lag) == _MAX_LAG_FRAMES

    verdict = "adequate" if abs(r) >= ADEQUATE_R_THRESHOLD else "inadequate"

    result = {
        "verdict": verdict,
        "pearson_r": round(r, 4),
        "best_lag_frames": lag,
        "lag0_pearson_r": round(lag0_r, 4),
        "fps": mouth["fps"],
        "n_frames": mouth["n_frames"],
        "n_detected": mouth["n_detected"],
        "mouth_ratio_mean": round(float(np.nanmean(ratios)), 4),
        "mouth_ratio_std": round(float(np.nanstd(ratios)), 4),
        "audio_rms_mean": round(float(np.mean(audio_env)), 6),
    }
    if lag_search_hit_boundary:
        result["caveat"] = (
            f"best_lag_frames hit the search boundary (±{_MAX_LAG_FRAMES}); on short "
            "clips this can reflect fewer valid sample pairs at extreme lags rather "
            "than genuine lip-sync offset — treat lag0_pearson_r as the more "
            "trustworthy statistic in this case."
        )
    return result


if __name__ == "__main__":
    import json
    import sys

    if len(sys.argv) != 2:
        print("usage: python -m app.lipsync_metrics <mp4_path>", file=sys.stderr)
        sys.exit(1)

    result = measure_lipsync_precision(sys.argv[1])
    print(json.dumps(result, indent=2))

File: swift/ltx-video-director/Sources/LTXVideoDirector/LipsyncMetrics.swift
//
//  LipsyncMetrics.swift
//  LTXVideoDirector
//
//  Native (Vision + AVFoundation) port of
//  python/mlx-movie-director/app/lipsync_metrics.py's
//  measure_lipsync_precision(): does generated mouth motion track the
//  audio, or is it "a face talking in general"? No Python, no mediapipe —
//  Vision's VNDetectFaceLandmarksRequest is a system framework, not a
//  ported model. Landmark topology differs from mediapipe's 468-point
//  FaceMesh, so this is a fresh implementation of the same CONCEPT
//  (mouth-gap / interocular-distance ratio, correlated against audio RMS
//  via a small lag search), not a numeric port — see
//  docs/superpowers/specs/2026-07-25-swift-lipsync-metrics-design.md for
//  the validation approach against the Python version's verdicts.
//

import AVFoundation
import Foundation
import Vision

public enum LipsyncMetrics {
    /// Resample `source` onto `count` points via linear interpolation over
    /// a shared [0, 1] parameterization of both series — same behavior as
    /// `np.interp(np.linspace(0,1,count), np.linspace(0,1,source.count), source)`.
    public static func linearResample(_ source: [Double], to count: Int) -> [Double] {
        guard count > 0 else { return [] }
        guard let first = source.first else { return Array(repeating: 0.0, count: count) }
        guard source.count > 1 else { return Array(repeating: first, count: count) }
        if count == 1 { return [source[0]] }
        var result: [Double] = []
        result.reserveCapacity(count)
        let srcLast = Double(source.count - 1)
        for i in 0..<count {
            let dstFrac = Double(i) / Double(count - 1)
            let srcPos = dstFrac * srcLast
            let lo = Int(srcPos.rounded(.down))
            let hi = min(lo + 1, source.count - 1)
            let frac = srcPos - Double(lo)
            result.append(source[lo] * (1 - frac) + source[hi] * frac)
        }
        return result
    }

    /// Best |Pearson r| between `a` and `b` over lags in [-maxLag, maxLag].
    /// Positive lag means `b` is shifted forward relative to `a` (`b` lags
    /// `a`) — matches the pairing convention of Python's `_lagged_pearson`
    /// in python/mlx-movie-director/app/lipsync_metrics.py, which this
    /// function ports. For `lag >= 0`, `a[lag...]` is paired against
    /// `b[0..<(n - lag)]`; for `lag < 0`, `a[0..<(n + lag)]` is paired
    /// against `b[(-lag)...]`. NaNs are dropped pairwise per-lag; a lag
    /// with fewer than 4 valid pairs is skipped. Returns (0.0, 0) if fewer
    /// than 8 total samples.
    ///
    /// The search order is ascending, `-maxLag...maxLag`, matching Python's
    /// `_lagged_pearson` exactly — a candidate must *strictly* improve on
    /// the best |r| seen so far to replace it, so on a tie the first-seen
    /// (most-negative) lag wins. This is a deliberate behavioral match, not
    /// just a numeric one: on ties, this biases toward the most-negative
    /// lag rather than the smallest |lag|, and callers relying on
    /// Swift-vs-Python lag agreement depend on that exact order.
    public static func laggedPearson(_ a: [Double], _ b: [Double], maxLag: Int) -> (r: Double, lag: Int) {
        let n = min(a.count, b.count)
        guard n >= 8 else { return (0.0, 0) }
        let a = Array(a.prefix(n))
        let b = Array(b.prefix(n))
        var bestR = 0.0
        var bestLag = 0
        for lag in -maxLag...maxLag {
            let aSeg: [Double]
            let bSeg: [Double]
            if lag >= 0 {
                guard lag < n else { continue }
                aSeg = Array(a[lag...])
                bSeg = Array(b[0..<(n - lag)])
            } else {
                guard -lag < n else { continue }
                aSeg = Array(a[0..<(n + lag)])
                bSeg = Array(b[(-lag)...])
            }
            guard let r = pearson(aSeg, bSeg), abs(r) > abs(bestR) else { continue }
            bestR = r
            bestLag = lag
        }
        return (bestR, bestLag)
    }

    /// Pearson correlation over paired samples, dropping any pair where
    /// either value is NaN. Returns nil if fewer than 4 valid pairs remain,
    /// or either series has ~zero variance.
    static func pearson(_ a: [Double], _ b: [Double]) -> Double? {
        var av: [Double] = []
        var bv: [Double] = []
        av.reserveCapacity(a.count)
        bv.reserveCapacity(a.count)
        for i in 0..<min(a.count, b.count) {
            if !a[i].isNaN && !b[i].isNaN {
                av.append(a[i])
                bv.append(b[i])
            }
        }
        guard av.count >= 4 else { return nil }
        let meanA = av.reduce(0, +) / Double(av.count)
        let meanB = bv.reduce(0, +) / Double(bv.count)
        let stdA = sqrt(av.reduce(0) { $0 + ($1 - meanA) * ($1 - meanA) } / Double(av.count))
        let stdB = sqrt(bv.reduce(0) { $0 + ($1 - meanB) * ($1 - meanB) } / Double(bv.count))
        guard stdA > 1e-9, stdB > 1e-9 else { return nil }
        var cov = 0.0
        for i in 0..<av.count { cov += (av[i] - meanA) * (bv[i] - meanB) }
        cov /= Double(av.count)
        return cov / (stdA * stdB)
    }

    /// r >= 0.3 (positive direction) is a conventional "weak-to-moderate but
    /// real" correlation floor — same value/rationale as the Python version's
    /// ADEQUATE_R_THRESHOLD. See Task 4 (validation against real clips) for
    /// whether this (and minMouthStdThreshold below) need retuning under
    /// Vision's landmark scale — ADEQUATE_R_THRESHOLD is a correlation
    /// coefficient and should be scale-invariant; minMouthStdThreshold is a
    /// raw ratio magnitude and is the one actually sensitive to
    /// landmark-scale differences from mediapipe.
    public static let adequateRThreshold = 0.3

    /// Same rationale as the Python version's MIN_MOUTH_STD_THRESHOLD: a
    /// near-flat mouth_ratio series can still clear adequateRThreshold by
    /// chance (Pearson r on a near-constant signal is noise-dominated).
    public static let minMouthStdThreshold = 0.01

    public struct VerdictResult {
        public let verdict: String
        public let caveat: String?
    }

    /// Direct port of measure_lipsync_precision's verdict/caveat decision
    /// tree. `lag0R` is accepted (not currently read) to keep the call site
    /// symmetric with the Python version's signature/data available at the
    /// call site — Swift's classify step does not use it to decide the
    /// verdict, only the lag-boundary caveat text references lag0R being
    /// "the more trustworthy statistic," same as Python.
    public static func classifyVerdict(r: Double, lag: Int, maxLag: Int, lag0R: Double, mouthRatioStd: Double) -> VerdictResult {
        let flatMouth = mouthRatioStd < minMouthStdThreshold
        let verdict = (r >= adequateRThreshold && !flatMouth) ? "adequate" : "inadequate"
        var caveat: String? = nil
        if flatMouth && r >= adequateRThreshold {
            caveat = "pearson_r cleared \(adequateRThreshold) but mouth_ratio_std "
                + "(\(String(format: "%.4f", mouthRatioStd))) is below \(minMouthStdThreshold) — the mouth "
                + "barely moves at all, so this correlation is almost certainly spurious "
                + "noise-alignment rather than real lip-sync. Treated as inadequate."
        } else if r <= -adequateRThreshold {
            caveat = "pearson_r (\(String(format: "%.4f", r))) is strongly negative — the mouth is anti-correlated "
                + "with audio loudness (opens when quiet, closes when loud), which is not "
                + "genuine lip-sync even though |r| clears threshold. Treated as inadequate."
        }
        // A plain `if` (not `else if`) — mirrors the Python version, where
        // the lag-boundary caveat OVERWRITES any caveat set above if both
        // conditions are true.
        if abs(lag) == maxLag {
            caveat = "best_lag_frames hit the search boundary (±\(maxLag)); on short "
                + "clips this can reflect fewer valid sample pairs at extreme lags rather "
                + "than genuine lip-sync offset — treat lag0_pearson_r as the more "
                + "trustworthy statistic in this case."
        }
        return VerdictResult(verdict: verdict, caveat: caveat)
    }

    public struct LipsyncResult: Codable {
        public let verdict: String
        public let pearsonR: Double?
        public let mouthRatioStd: Double?
        public let caveat: String?
        public let note: String?
        public let bestLagFrames: Int?
        public let lag0PearsonR: Double?
        public let fps: Double?
        public let nFrames: Int?
        public let nDetected: Int?
        public let mouthRatioMean: Double?
        public let audioRmsMean: Double?

        enum CodingKeys: String, CodingKey {
            case verdict
            case pearsonR = "pearson_r"
            case mouthRatioStd = "mouth_ratio_std"
            case caveat, note
            case bestLagFrames = "best_lag_frames"
            case lag0PearsonR = "lag0_pearson_r"
            case fps
            case nFrames = "n_frames"
            case nDetected = "n_detected"
            case mouthRatioMean = "mouth_ratio_mean"
            case audioRmsMean = "audio_rms_mean"
        }
    }

    struct MouthSeries {
        let ratios: [Double]  // NaN for frames with no detected face
        let fps: Double
        let nFrames: Int
        let nDetected: Int
    }

    /// Lag search window in video frames — same value as the Python
    /// version's _MAX_LAG_FRAMES.
    public static let maxLagFrames = 4

    static func mean(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0.0 }
        return values.reduce(0, +) / Double(values.count)
    }

    static func standardDeviation(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0.0 }
        let m = mean(values)
        let variance = values.reduce(0) { $0 + ($1 - m) * ($1 - m) } / Double(values.count)
        return sqrt(variance)
    }

    private static func centroid(_ points: [CGPoint]) -> CGPoint {
        guard !points.isEmpty else { return .zero }
        let sum = points.reduce(CGPoint.zero) { CGPoint(x: $0.x + $1.x, y: $0.y + $1.y) }
        return CGPoint(x: sum.x / Double(points.count), y: sum.y / Double(points.count))
    }

    /// Inner-lip vertical extent / interocular distance, in image-pixel
    /// space, for one face observation. `boundingBox` is Vision's
    /// full-image-normalized, bottom-left-origin face rect;
    /// `landmarks.*.normalizedPoints` are normalized WITHIN that box —
    /// converting both to a shared pixel space cancels out face-scale/zoom,
    /// same intent as the Python version's mediapipe-normalized-space ratio.
    static func mouthOpenRatio(landmarks: VNFaceLandmarks2D, boundingBox: CGRect, imageWidth: Int, imageHeight: Int) -> Double? {
        guard let innerLips = landmarks.innerLips, innerLips.pointCount > 0,
              let leftEye = landmarks.leftEye, leftEye.pointCount > 0,
              let rightEye = landmarks.rightEye, rightEye.pointCount > 0
        else { return nil }

        func toPixel(_ p: CGPoint) -> CGPoint {
            CGPoint(
                x: (boundingBox.origin.x + p.x * boundingBox.width) * Double(imageWidth),
                y: (boundingBox.origin.y + p.y * boundingBox.height) * Double(imageHeight)
            )
        }

        let lipYs = innerLips.normalizedPoints.map { toPixel($0).y }
        guard let maxY = lipYs.max(), let minY = lipYs.min() else { return nil }
        let mouthGap = maxY - minY

        let leftCenter = centroid(leftEye.normalizedPoints.map(toPixel))
        let rightCenter = centroid(rightEye.normalizedPoints.map(toPixel))
        let interocular = hypot(leftCenter.x - rightCenter.x, leftCenter.y - rightCenter.y)
        guard interocular > 1e-6 else { return nil }
        return mouthGap / interocular
    }

    /// Per-frame mouth-open ratio for every frame of the video (NaN where no
    /// face is detected). Reuses VideoProbe's existing consecutive-frame
    /// extraction (the same utility VideoGate's motion check uses) rather
    /// than writing a new AVAssetReader frame walk.
    ///
    /// Scope: this decodes and holds EVERY frame of the clip in memory at
    /// once (via `VideoProbe.consecutiveFrames`) and runs a Vision face
    /// request per frame — fine for this repo's actual use case (short
    /// 2-5s talking-head dialogue shots, per the design doc), but would be
    /// a real memory/perf concern on long clips. Not a general-purpose
    /// long-video tool — don't reach for this on a 10-minute video without
    /// budgeting for the cost.
    ///
    /// Assumes `VideoProbe.consecutiveFrames` returns frames without gaps
    /// — it silently drops (does not pad) any frame it fails to decode, so
    /// a dropped mid-clip frame would shift every later position by one,
    /// desyncing the array-position-as-time-proxy this series is built on
    /// (the audio envelope gets resampled onto these same positions). In
    /// practice this is low-risk: the ±`maxLagFrames` lag search absorbs
    /// small timing skews like this, and dropped frames should be rare on
    /// clean mp4s — but it's not a guarantee of perfect frame-to-index
    /// correspondence.
    static func extractMouthOpenSeries(url: URL) throws -> MouthSeries {
        let info = try VideoProbe.info(url: url)
        guard info.fps > 0, info.duration > 0 else {
            return MouthSeries(ratios: [], fps: info.fps, nFrames: 0, nDetected: 0)
        }
        let frameCount = max(1, Int((info.duration * info.fps).rounded()))
        let frames = try VideoProbe.consecutiveFrames(url: url, startTime: 0, count: frameCount, fps: info.fps)

        var ratios: [Double] = []
        ratios.reserveCapacity(frames.count)
        var nDetected = 0
        for image in frames {
            var ratio = Double.nan
            let request = VNDetectFaceLandmarksRequest()
            let handler = VNImageRequestHandler(cgImage: image, options: [:])
            if (try? handler.perform([request])) != nil,
               let face = request.results?.first,
               let landmarks = face.landmarks,
               let r = mouthOpenRatio(landmarks: landmarks, boundingBox: face.boundingBox, imageWidth: image.width, imageHeight: image.height) {
                ratio = r
                nDetected += 1
            }
            ratios.append(ratio)
        }
        return MouthSeries(ratios: ratios, fps: info.fps, nFrames: frames.count, nDetected: nDetected)
    }

    /// Decode the first audio track to mono Float32 PCM. Deliberately
    /// separate from AudioProbe.swift's own private decode loop (same
    /// AVAssetReader settings, small duplication) rather than refactoring
    /// AudioProbe to expose raw samples — that file serves a different
    /// consumer (VideoGate) and isn't part of this change's scope.
    private static func decodeMonoPCM(url: URL) -> [Float] {
        let asset = AVURLAsset(url: url)
        guard let track = asset.tracks(withMediaType: .audio).first,
              let reader = try? AVAssetReader(asset: asset)
        else { return [] }
        let settings: [String: Any] = [
            AVFormatIDKey: kAudioFormatLinearPCM,
            AVLinearPCMBitDepthKey: 32,
            AVLinearPCMIsFloatKey: true,
            AVLinearPCMIsNonInterleaved: false,
            AVNumberOfChannelsKey: 1,
        ]
        let output = AVAssetReaderTrackOutput(track: track, outputSettings: settings)
        reader.add(output)
        reader.startReading()
        var samples: [Float] = []
        while let buffer = output.copyNextSampleBuffer() {
            guard let blockBuffer = CMSampleBufferGetDataBuffer(buffer) else { continue }
            let length = CMBlockBufferGetDataLength(blockBuffer)
            var data = [UInt8](repeating: 0, count: length)
            CMBlockBufferCopyDataBytes(blockBuffer, atOffset: 0, dataLength: length, destination: &data)
            data.withUnsafeBytes { raw in
                let floats = raw.bindMemory(to: Float32.self)
                samples.append(contentsOf: floats)
            }
        }
        return samples
    }

    /// Audio RMS envelope resampled onto `nSamples` points spanning the
    /// clip — mirrors the Python version's extract_audio_envelope.
    static func extractAudioEnvelope(url: URL, nSamples: Int) -> [Double] {
        guard nSamples > 0 else { return [] }
        let samples = decodeMonoPCM(url: url)
        guard samples.count >= 1024 else { return Array(repeating: 0.0, count: nSamples) }

        let windowSize = 1024
        var rmsFrames: [Double] = []
        var i = 0
        while i < samples.count {
            let end = min(i + windowSize, samples.count)
            let chunk = samples[i..<end]
            var sumSq: Float = 0
            for v in chunk { sumSq += v * v }
            rmsFrames.append(Double(sqrt(sumSq / Float(chunk.count))))
            i = end
        }
        guard rmsFrames.count >= 2 else { return Array(repeating: 0.0, count: nSamples) }
        return linearResample(rmsFrames, to: nSamples)
    }

    /// End-to-end: extract both series, correlate, verdict. Mirrors
    /// measure_lipsync_precision()'s structure exactly (same early-exit
    /// order: no_face check before no_audio check).
    ///
    /// Divergence from Python: `measure_lipsync_precision` never raises —
    /// it always returns a verdict dict, even for a totally undecodable
    /// file. This port intentionally DOES throw for a genuinely unreadable
    /// file (via `VideoProbe.info`'s `VideoProbeError.noVideoTrack` etc.),
    /// matching how other commands in this package already handle this
    /// (e.g. `VideoGate.evaluate` also throws) — the CLI command layer
    /// lets ArgumentParser report the thrown error with a nonzero exit
    /// rather than this function synthesizing an error verdict itself.
    public static func measure(url: URL) throws -> LipsyncResult {
        let mouth = try extractMouthOpenSeries(url: url)
        guard mouth.nDetected >= 4 else {
            return LipsyncResult(
                verdict: "no_face", pearsonR: nil, mouthRatioStd: nil, caveat: nil,
                note: "insufficient face detections", bestLagFrames: nil, lag0PearsonR: nil,
                fps: mouth.fps, nFrames: mouth.nFrames, nDetected: mouth.nDetected,
                mouthRatioMean: nil, audioRmsMean: nil)
        }

        let audioEnv = extractAudioEnvelope(url: url, nSamples: mouth.ratios.count)
        let audioStd = standardDeviation(audioEnv.filter { !$0.isNaN })
        guard audioStd >= 1e-9 else {
            return LipsyncResult(
                verdict: "no_audio", pearsonR: nil, mouthRatioStd: nil, caveat: nil,
                note: "audio envelope has ~zero variance (silent/missing track)",
                bestLagFrames: nil, lag0PearsonR: nil,
                fps: mouth.fps, nFrames: mouth.nFrames, nDetected: mouth.nDetected,
                mouthRatioMean: nil, audioRmsMean: nil)
        }

        let (r, lag) = laggedPearson(mouth.ratios, audioEnv, maxLag: maxLagFrames)
        let lag0R = pearson(mouth.ratios, audioEnv) ?? 0.0
        let validRatios = mouth.ratios.filter { !$0.isNaN }
        let mouthRatioStd = standardDeviation(validRatios)
        let mouthRatioMean = mean(validRatios)
        let audioRmsMean = mean(audioEnv)

        let classification = classifyVerdict(r: r, lag: lag, maxLag: maxLagFrames, lag0R: lag0R, mouthRatioStd: mouthRatioStd)

        func round4(_ v: Double) -> Double { (v * 10000).rounded() / 10000 }
        func round6(_ v: Double) -> Double { (v * 1_000_000).rounded() / 1_000_000 }

        return LipsyncResult(
            verdict: classification.verdict,
            pearsonR: round4(r),
            mouthRatioStd: round4(mouthRatioStd),
            caveat: classification.caveat,
            note: nil,
            bestLagFrames: lag,
            lag0PearsonR: round4(lag0R),
            fps: mouth.fps,
            nFrames: mouth.nFrames,
            nDetected: mouth.nDetected,
            mouthRatioMean: round4(mouthRatioMean),
            audioRmsMean: round6(audioRmsMean)
        )
    }
}
